### DISCUSSION TITLE
How to apply a constraint over an immersed boundary?

### DISCUSSION THREAD

**Discussion by doquang** [idaholab/moose#23624] [Q&A General] [answered]
Hi, 

I would like to know is there any ways in Moose/Libmesh that I can use to apply a constraint into an immersed boundary?

Thanks a lot,
/ Minh

**Comment by GiudGiud** [comment]
Hello

In which context is this? Which physics?

Guillaume

**Reply by doquang** [reply]
We would like to apply a wall boundary for fluid flow and a fixed temperature on grid points that do not conform to the shape of the boundaries.

**Reply by GiudGiud** [reply]
We may be able to do that with the constraints system by evaluating the shape functions of the variables at an arbitrary location rather than on the boundary. 

We may be able to leverage the functor system to do that, with an ElemPoint argument (evaluates the variable at a point in an element)
https://mooseframework.inl.gov/docs/doxygen/moose/classMoose_1_1FunctorBase.html#adf3479c9e85154a9d81b59cd77926b5b

or a generic projector but it's likely more complicated here

@roystgnr @lindsayad for ideas

**Reply by lindsayad** [reply]
We can't currently do this with the MOOSE constraint system. I'll leave @roystgnr to comment on the libMesh side of things although I believe he's currently on personal leave

**Reply by hugary1995** [reply]
How do you want to represent the immersed boundary?

**Reply by roystgnr** [reply, marked as answer]
libMesh will let you apply an arbitrary linear (homogeneous or heterogeneous) constraint equation to any DoF you want, for as many DoFs as you want, but it won't let you apply multiple constraints to the same DoF and you always have to choose which DoF is "the" constrained one, which makes that trick hard to reuse in complex codes, making it a last-resort feature to use in practice.  In your case the "choose which DoF" would be the tricky part.

Penalty methods, Nitsche-type methods, lagrange multiplier methods, etc. might be more flexible.  Even in the super-simple case of "pressure pinning", where you have a constant-pressure kernel to constrain away, I've always preferred adding a penalty term to the residual (see example fem_system_ex1 in libMesh) and others prefer a Lagrange multiplier (see example systems_of_equations_ex3 there).
